### PR TITLE
Forbid partial data constructors exports

### PR DESCRIPTION
**Pull request by kl0tl**
It seemed appropriate to reuse the `TransitiveDctorExportError` error. I stored the constructors names under a list for consistency with `TransitiveExportError` but perhaps non empty lists would be better 🤷‍♂️ 

Close https://github.com/purescript/purescript/issues/3869.

### CODE REVIEW

File: src/Language/PureScript/TypeChecker.hs [hunk 1]
```diff
@@ -712,3 +713,19 @@ typeCheckModule (Module ss coms mn decls (Just exps)) =
     extractMemberName (TypeDeclaration td) = tydeclIdent td
     extractMemberName _ = internalError "Unexpected declaration in typeclass member list"
   checkClassMembersAreExported _ = return ()
+
+  checkDataConstructorsAreExported :: DeclarationRef -> m ()
+  checkDataConstructorsAreExported dr@(TypeRef ss' name (Just exportedDataConstructorsNames))
+    | not (null exportedDataConstructorsNames) = do
+      let dataConstructorNames = fromMaybe [] . headMay $ mapMaybe findDataConstructorsNames decls
```
Comment (natefaubion) [inline src/Language/PureScript/TypeChecker.hs:720]: Data constructors are all in the environment, so we shouldn't need to filter through all declarations for every data export. They are under `types`, and part of the unfortunately named `TypeKind` slot.

File: src/Language/PureScript/TypeChecker.hs [hunk 2]
```diff
@@ -712,3 +713,18 @@ typeCheckModule (Module ss coms mn decls (Just exps)) =
     extractMemberName (TypeDeclaration td) = tydeclIdent td
     extractMemberName _ = internalError "Unexpected declaration in typeclass member list"
   checkClassMembersAreExported _ = return ()
+
+  checkDataConstructorsAreExported :: DeclarationRef -> m ()
```
Comment (hdgarrood) [inline src/Language/PureScript/TypeChecker.hs:717]: Would you mind adding a comment here? I think something along the lines of "if any data constructors are exported, we require all data constructors to be exported" would be enough.

**Review by hdgarrood — APPROVED** [review]
Looks great, thanks!

File: src/Language/PureScript/TypeChecker.hs (hunk 2 above)
Comment (kl0tl) [reply src/Language/PureScript/TypeChecker.hs:718]: Sure! I added a comment.

**Pull request merged**